website/blog/2025-07-01-mlflow-prompt-evaluate/index.md

```diff
@@ -0,0 +1,435 @@
+            else:
+                flat_items = contents
+            return _extract_qa_pairs(flat_items)
+        else:
+            return _extract_qa_pairs(contents)
+
+```
+
+What does _\_extract_qa_pairs_ do here? Let's break down it down for a second:
+
+- We create a look-up dictionary _qa_pairs_
+- We identify "question" items that have linked answer pair(s) in the form of _(q_id, a_id)_
+- Construct question-answer pairs by joining the individual words  
+  <br/>
+  Subsequently, we can call _get_json_ to get the OCR-identified question/answer structure based on the form image data.
+
+```python
+
+from utils import get_image, get_json, get_random_files
+from IPython.display import Image, display
+
+random_file = get_random_files()
+image_bytes = get_image(random_file)
+get_json(random_file)
+
+```
+
+#### 2. Example LLM Call
+
+Before we make a call to OpenAI, we need to set up MLflow tracking and create and experiment. We can then automatically log all the API calls we make using `autolog()`
+
+```python
+import mlflow
+
+mlflow.set_tracking_uri(os.getcwd() + "/mlruns")
+mlflow.set_experiment("quickstart")
+mlflow.openai.autolog()
+```
+
+Our next step will be to set mlruns as the local directory for MLflow tracking. We will create an experiment called quickstart and log the API calls for metrics like etc.
+
+On the OpenAI side, we initialize the OpenAI client and provide the prompt to the LLM (for example, instructing the LLM to act as an OCR expert).
+
+```python
+client = OpenAI()
+
+system_prompt = """You are an expert at Optical Character Recognition (OCR). Extract the questions and answers from the image."""
+base64_image = get_image(random_file, encode_as_str=True)
+```
+
+_get_image()_ can be a neat helper function that can handle
+
+        Reading the image file from a specified path
+        Resizing the file and convert it into a JPEG of a specified quality
+
+```python
+import base64
+from PIL import Image
+from io import BytesIO
+
+def _compress_image(file_path: str, quality: int = 40, max_size: tuple[int, int] = (1000, 1000)) -> bytes:
+    """Compresses an image by resizing and converting to JPEG with given quality."""
+    with Image.open(file_path) as img:
+        img = img.convert("RGB")  # Ensure JPEG compatibility
+        img.thumbnail(max_size)  # Resize
+        buf = BytesIO()
+        img.save(buf, format="JPEG", quality=quality)
+        return buf.getvalue()
+
+
+def get_image(
+    file_name: str, directory: str = IMAGES_DIRECTORY, encode_as_str: bool = False
+) -> bytes:
+    file_name += ".png" if not file_name.endswith(".png") else file_name
+    path = os.path.join(directory, file_name)
+    with open(path, "rb") as f:
+        file = f.read()
+        compressed = _compress_image(path)
+
+        if encode_as_str:
+            return base64.b64encode(compressed).decode("utf-8")
+        else:
+            return compressed
+```
+
+#### 3. Tracing UI [WIP]
+
```

Comment:
Add a paragraph or two describing tracing and why we care. It's a flagship product, so we want to focus on it's value prop